RHDEVDOCS-6488: creating build with buildpacks (#95748)
openshift/openshift-docs: modules/ob-creating-a-buildpacks-build.adoc

* `buildpacks` strategy
* `buildpacks-extender` strategy

[NOTE]
====
The `buildpacks-extender` strategy is compatible with the experimental buildpacks `extender` lifecycle phase. You can use this strategy e.g for builds based on Node.js. However, the `extender` phase does not support Quarkus or any other Java-based builds. For such builds, you should use the `buildpacks` strategy instead.
====

[IMPORTANT]
====
Buildpacks is a Technology Preview feature only. Technology Preview features are not supported with Red{nbsp}Hat production service level agreements (SLAs) and might not be functionally complete. Red{nbsp}Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red{nbsp}Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Support Scope].
====

.Prerequisites

* You have installed the {builds-operator} on the {ocp-product-title} cluster.
* You have installed the `oc` CLI.

> * You have created the project where your final application image will be stored by using the command `oc new-project buildpacks-example`.

Reviewer: ```suggestion
* You have created the project where your final application image is stored by using the command `oc new-project buildpacks-example`.
```
Try to avoid future tense.

* Optional: You have installed the `shp` CLI.

[IMPORTANT]
====

> Using `shp` CLI with buildpacks requires additional permissions setup

Reviewer: ```suggestion
Using `shp` CLI with buildpacks requires additional permissions setup that
```

Reviewer: It looks like you have a hard return between line 33 and 34. Please correct this.

you must complete before you start creating a buildpacks build.
====

.Procedure

> . OPTIONAL: To use the `shp` CLI with buildpacks, you must first grant permission to the `pipeline` service to access the image registry in the `buildpacks-example` project  with the following two commands:

Reviewer 2: ```suggestion
. OPTIONAL: To use the `shp` CLI with buildpacks, you must first grant permission to the `pipeline` service account to access the image registry in the `buildpacks-example` project  with the following two commands:
```

+
[source,terminal]
----
$ oc create rolebinding allow-builds-to-push \

  --clusterrole=edit \

 --serviceaccount=default:pipeline \

--namespace=buildpacks-example
----
. Create a `Build` resource and apply it to the {ocp-product-title} cluster:
.. Using `oc` CLI
+
[source,terminal]
----
$ oc apply -f - <<EOF
apiVersion: shipwright.io/v1beta1

kind: Build

metadata:

  name: buildpack-nodejs-build

spec:

  source: <1>

    type: Git

    git:

      url: https://github.com/redhat-openshift-builds/samples.git

  strategy: <2>

    name: buildpacks

    kind: ClusterBuildStrategy

  retention:

    atBuildDeletion: true

  paramValues: <3>

    - name: run-image <4>

      value: paketobuildpacks/run-ubi8-base:latest

    - name: cnb-builder-image <5>

      value: paketobuildpacks/builder-jammy-tiny:0.0.344

    - name: source-subpath <6>

      value: "buildpacks/nodejs"

  output: <7>
    image: image-registry.openshift-image-registry.svc:5000/buildpacks-example/taxi-app
EOF
----
<1> The Git repository containing your application source code.
<2> The build strategy to build the container.
<3> Parameters set for the buildpacks strategy.

> <4> The base image on which your application will run.

Reviewer: ```suggestion
<4> The base image on which your application runs.
```

----
[NOTE]
====
The `BuildRun` resource creates a `TaskRun` resource, which then creates the pods to execute build strategy steps.
====

.Verification
. Wait for all containers to complete their tasks.
. Check if the pod shows the `STATUS` field as `Completed`:
+
[source,terminal]
----
$ oc get pods -w
----
. Check if the `TaskRun` resource shows the `SUCCESS` field as `True`:
+
[source,terminal]
----
$ oc get tr
----
. Check if the `BuildRun` resource shows the `SUCCESS` field as `True`:
+
[source,terminal]
----
$ oc get br
----
+
[NOTE]
====
If the build run fails, you can check the `status.failureDetails` field in your `BuildRun` resource to identify the exact point where the failure happened in the pod or container.

The pod might switch to a `NotReady` state because one of the containers has completed its task. This is an expected behavior.
====
. Check if the image has been pushed to the registry you specified in the `build.spec.output.image` field by running the following command from a node that can access the internal registry to pull the image:
+
[source,terminal]
----
$ podman pull
----
+
[NOTE]
====
An example output of the `podman pull` command might look like this:

> 

Reviewer: Add heading here, such as: ".Example output"

Author: done!
